core/src/main/java/org/infinispan/remoting/transport/jgroups/CommandAwareRpcDispatcher.java

```diff
@@ -401,50 +379,45 @@ private void staggeredProcessNext(ReplicableCommand command, RspFilter filter, L
          return;
       }
       try {
-         processCallsStaggered(command, filter, dests, mode, deliverOrder, marshaller, theFuture,
-               destIndex + 1, deadline, rsps);
+         processCallsStaggered(command, filter, dests, mode, deliverOrder, theFuture, destIndex + 1, deadline, rsps);
       } catch (Exception e) {
          // We should never get here, any remote exception will be in the Rsp
          theFuture.completeExceptionally(e);
       }
    }
 
-   private RspListFuture processCalls(ReplicableCommand command, boolean broadcast, long timeout,
-                                      RspFilter filter, List<Address> dests, ResponseMode mode,
-                                      DeliverOrder deliverOrder, Marshaller marshaller) throws Exception {
+   private RspListFuture processCalls(ReplicableCommand command, boolean broadcast, long timeout, RspFilter filter,
+         List<Address> dests, ResponseMode mode, DeliverOrder deliverOrder) throws Exception {
       if (trace) log.tracef("Replication task sending %s to addresses %s with response mode %s", command, dests, mode);
       boolean rsvp = isRsvpCommand(command);
 
-      Buffer buf = marshallCall(marshaller, command);
-      Message msg;
-      RequestOptions opts;
+      Buffer buf = marshallCall(command);
+      RequestOptions opts = constructRequestOptions(mode, rsvp, deliverOrder, timeout);
+      Collection<Address> realDest = dests;
       if (deliverOrder == DeliverOrder.TOTAL) {
-         msg = constructMessage(buf, new AnycastAddress(dests), mode, rsvp, deliverOrder);
-         opts = new RequestOptions(mode, timeout, false, filter);
+         opts.anycasting(true).useAnycastAddresses(true);
       } else if (broadcast || FORCE_MCAST) {
-         msg = constructMessage(buf, null, mode, rsvp, deliverOrder);
-         opts = new RequestOptions(mode, timeout, false, filter);
+         opts.anycasting(false);
+         realDest = null; //broadcast
       } else {
-         msg = constructMessage(buf, null, mode, rsvp, deliverOrder);
-         opts = new RequestOptions(mode, timeout, true, filter);
+         opts.anycasting(true).setUseAnycastAddresses(false);
       }
+      opts.rspFilter(filter);
 
-      GroupRequest<Response> request = cast(dests, msg, opts, false);
-      if (mode == ResponseMode.GET_NONE)
-         return null;
-
+      GroupRequest<Response> request = cast(realDest, buf, opts, false);
+      if (mode == ResponseMode.GET_NONE) return null;
       RspListFuture retval = new RspListFuture(request);
       if (request == null) {
```

Comment:
I don't think the request can be ```null```. Even if it is, the ```RspListFuture``` constructor will throw a NPE.